lib/buffer.js

```diff
@@ -304,6 +306,20 @@ function slowToString(encoding, start, end) {
   if (end > this.length) end = this.length;
   if (end <= start) return '';
 
+  switch (encoding) {
+    case 'ucs2':
+    case 'ucs-2':
+    case 'utf16le':
+    case 'utf-16le':
+      if (((end - start) / 2) > kStringMaxLength)
+        throw new RangeError('invalid String length');
```

Comment:
The division by two is because UCS-2 and UTF-16 are two-byte encodings (i.e. each character is two bytes) and V8's `String::kMaxLength` is in bytes.